File: 백준/Silver/20364. 부동산 다툼/부동산 다툼.java
import java.io.*;
import java.util.*;

public class Main {




    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n=Integer.parseInt(st.nextToken());
        int q=Integer.parseInt(st.nextToken());

        boolean[] visited=new boolean[n+1];

        for(int i=0;i<q;i++){
            int key=Integer.parseInt(br.readLine());
            int num=key;
            Stack<Integer> stack=new Stack<>();
            boolean can=true;

            while(num>1){
                stack.push(num);
                num/=2;
            }
            while(!stack.isEmpty()){
                int targ=stack.pop();
                if(visited[targ]){
                    System.out.println(targ);
                    can=false;
                    break;
                }

            }
            if(can){
                System.out.println(0);
                visited[key]=true;
            }
        }

    }}

File: 백준/Silver/20364. 부동산 다툼/부동산 다툼.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>

using namespace std;


int main(){

    ios_base::sync_with_stdio(0);
    cin.tie(0);


    int n;
    cin>>n;
    int q;
    cin>>q;

    vector<bool> visited(n+1,false);

    for(int i=0;i<q;i++){
        int num;
        cin>>num;
        int key=num;
        stack<int>st;
        bool canreach=true;
        while(key>1){
            st.push(key);
            key/=2;
        }
        while(!st.empty()){
            int targ=st.top();
            if(visited[targ]){
                cout<<targ<<'\n';
                canreach=false;
                break;
            }
            else
                st.pop();
        }
        if (canreach){
            cout<<0<<'\n';
            visited[num]=true;
        }


    }
    return 0;
}

